pkg/api/operators/operators_test.go

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) 2020, Oracle and/or its affiliates.
+// Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+package operators
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/gorilla/mux"
+	"github.com/stretchr/testify/assert"
+	"github.com/verrazzano/verrazzano-operator/pkg/controller"
+	"github.com/verrazzano/verrazzano-operator/pkg/testutil"
+	"github.com/verrazzano/verrazzano-operator/pkg/testutilcontroller"
+	"github.com/verrazzano/verrazzano-operator/pkg/types"
+	v1wlsopr "github.com/verrazzano/verrazzano-wko-operator/pkg/apis/verrazzano/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/fake"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+const wlsNamespace, wlsOperatorName, wlsBindingName = "test", "wls-operator", "mytestbinding"
+
+// TestInit ensures that a set of operators can be obtained from the provided set of listers.
+func TestInit(t *testing.T) {
+	// GIVEN an array of clusters and a model binding pair is provided
+	Operators = nil
+	clusters := testutil.GetTestClusters()
+	var clients kubernetes.Interface = fake.NewSimpleClientset()
+	modelBindingPairs := map[string]*types.ModelBindingPair{
+		"test-pair-1": testutil.GetModelBindingPair(),
+	}
+	type args struct {
+		listers controller.Listers
+	}
+	tests := []struct {
+		name    string
+		listers controller.Listers
+	}{
+		{name: "testInit",
+			listers: testutilcontroller.NewControllerListers(&clients, clusters, &modelBindingPairs)},
+	}
+	for _, tt := range tests {
+		// add a WLSOperator to one of the clusters
+		t.Run(tt.name, func(t *testing.T) {
+			// WHEN a WLS operator is appended to the cluster and the Init() method is invoked
+			createAndInsertWLSOperator(1, tt)
+			// ThEN the single operator with the expected name will be found in the Operators slice
+			assert.Equal(t, tt.listers, listerSet, "Wrong number of listers")
+			assert.Equal(t, 1, len(Operators), "Wrong number of WLS operators")
+			assert.Equal(t, "wls-operator-mytestbinding-1", Operators[0].Name, "Operator has the wrong name")
+		})
+	}
+}
+
+//TestReturnNoOperators ensures that an empty response is returned if no operators are available
+func TestReturnNoOperators(t *testing.T) {
+	// GIVEN an empty model binding pair is provided
+	Operators = nil
+	emptyModelBindingPairs := make(map[string]*types.ModelBindingPair)
+	Init(controller.Listers{
+		ModelBindingPairs: &emptyModelBindingPairs,
+	})
+	// setup http request processing
+	recorder := httptest.NewRecorder()
+	router := mux.NewRouter()
```

Comment:
what is this testing? this is not using our code right, it is using mux?

Reply:
Yes.  See api packages for multiple usages of this pattern in testing.

Reply:
To be clear - this code registers a HTTP handler that forwards the requests to our code.  For example, an invocation of "/operators/1" will forward the request to the ReturnSingleOperator:

			router.HandleFunc("/operators/{id}", ReturnSingleOperator)

Reply:
ok, i see now, i missed that.  as you say, even this feels a bit too much for a unit test.

Reply:
I'm not sure I completely agree.  IMO this is similar to the use of fakes/mocks etc.  The use of a lightweight HTTP framework to test the interactions this class supports seems valid.  In fact, the use of this approach unearthed an issue with the code - single operator responses for non-existent operators were returning an HTTP 200 response rather than a 404 response.  Given the nature of the API (support of REST interactions) the testing of the methods with a more lightweight HTTP framework that is more readily utilized as a testing infrastructure seems valid.

Reply:
I had some initial qualms about this as well because it seemed to be in-between a unit and a functional test, but for the reasons Jon mentioned, I think using this lightweight HTTP framework is a reasonable idea.

Reply:
Oh, also...there is a slightly more "unit-test-ish" way of testing these, which I did not think of until I saw Doug use it somewhere yesterday. The "recorder" satisfies the interface of a http response, so you could directly invoke the handler function and assert on the recorder, instead of using router and servehttp. That might be the compromise way. That will allow us to test status codes as wel.